cbits/win32/runProcess.c

```diff
@@ -0,0 +1,611 @@
+                                 _stdhandle,
+                                 GetCurrentProcess(), tmpHandle,
+                                 0,
+                                 TRUE, /* inheritable */
+                                 DUPLICATE_SAME_ACCESS);
+        if (!status) return false;
+        *destination = *tmpHandle;
+    }
+
+    return true;
+}
+
+/* Common functionality between the Posix FD version and native HANDLE version
+   of runInteractiveProcess.  The main difference lies in the use of
+   ASYNCHRONOUS which indicates whether the pipes that are created allow for
+   asynchronous access or not.  */
+
+static ProcHandle
+runInteractiveProcessWrapper (
+    wchar_t *cmd, wchar_t *workingDirectory,
+    wchar_t *environment,
+    HANDLE _stdin, HANDLE _stdout, HANDLE _stderr,
+    HANDLE *pStdInput, HANDLE *pStdOutput, HANDLE *pStdError,
+    int flags, bool useJobObject, HANDLE *hJob, bool asynchronous)
+{
+    STARTUPINFO sInfo;
+    PROCESS_INFORMATION pInfo;
+    HANDLE hStdInputRead   = INVALID_HANDLE_VALUE;
+    HANDLE hStdInputWrite  = INVALID_HANDLE_VALUE;
+    HANDLE hStdOutputRead  = INVALID_HANDLE_VALUE;
+    HANDLE hStdOutputWrite = INVALID_HANDLE_VALUE;
+    HANDLE hStdErrorRead   = INVALID_HANDLE_VALUE;
+    HANDLE hStdErrorWrite  = INVALID_HANDLE_VALUE;
+    BOOL close_fds = ((flags & RUN_PROCESS_IN_CLOSE_FDS) != 0);
+    // We always pass a wide environment block, so we MUST set this flag
+    DWORD dwFlags = CREATE_UNICODE_ENVIRONMENT;
+    BOOL inherit;
+
+    ZeroMemory(&sInfo, sizeof(sInfo));
+    sInfo.cb = sizeof(sInfo);
+    sInfo.dwFlags = STARTF_USESTDHANDLES;
+    ZeroMemory(&pInfo, sizeof(pInfo));
+
+    HANDLE defaultStdIn     = GetStdHandle(STD_INPUT_HANDLE);
+    HANDLE defaultStdOutput = GetStdHandle(STD_OUTPUT_HANDLE);
+    HANDLE defaultStdError  = GetStdHandle(STD_ERROR_HANDLE);
+
+    if (!setStdHandleInfo (&sInfo.hStdInput, _stdin, &hStdInputRead,
+                           &hStdInputWrite, defaultStdIn, TRUE, FALSE,
+                           asynchronous))
+      goto cleanup_err;
+
+    if (!setStdHandleInfo (&sInfo.hStdOutput, _stdout, &hStdOutputRead,
+                           &hStdOutputWrite, defaultStdOutput, FALSE, TRUE,
+                           asynchronous))
+      goto cleanup_err;
+
+    if (!setStdHandleInfo (&sInfo.hStdError, _stderr, &hStdErrorRead,
+                           &hStdErrorWrite, defaultStdError, TRUE, TRUE,
+                           asynchronous))
+      goto cleanup_err;
+
+    if (sInfo.hStdInput     !=  defaultStdIn
+        && sInfo.hStdOutput !=  defaultStdOutput
+        && sInfo.hStdError  !=  defaultStdError
+        && (flags & RUN_PROCESS_IN_NEW_GROUP) == 0)
+            dwFlags |= CREATE_NO_WINDOW;   // Run without console window only when both output and error are redirected
+
+    // See #3231
+    if (close_fds
+        && _stdin == defaultStdIn
+        && _stdout == defaultStdOutput
+        && _stderr == defaultStdError) {
```

Comment:
Should it not be an error if the user sets `close_fds` yet then asks us to redirect a standard handle? I'm worried that the user might be thinking that `close_fds` gives them security which they aren't in fact getting.

Reply:
I agree that this carve-out is weird, though in this case I only refactored the code and made it easier to read but left the same logic in place.  In particular this code seems to have been added as a workaround for https://gitlab.haskell.org/ghc/ghc/issues/3231 which is the thing we were still struggling with until recently. 

 The core premise being they have to wait for all the resources to be released by the process before continuing, however the issue was they waited for the main process to terminate and not all process that have inherited the handle.  This caused the same issues we were seeing with the race conditions in process spawning (just 11 years ago). 

The fix that was implemented was to not make handles inheritable if they aren't redirected. which means when the main process exits the handles are done.  This sorta works but only when you don't redirect the handles, either yourself or from the console.  It doesn't take into account when your handles are actually files and not pipes.  Which is why the bug was re-opened years ago.

This brings back to your point that maybe process jobs should be the default, after this is all said and done, we should evaluate that and make sure our current fix covers this case and we can remove this carve out.

I have created https://github.com/haskell/process/issues/186 to re-evaluate this fix, but for now I think it's easier if this PR doesn't change functionality.

Reply:
Alright, fair enough. This seems reasonable. Thanks for opening the ticket.